core/src/main/java/org/apache/iceberg/formats/ReadBuilder.java

```diff
@@ -0,0 +1,118 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.formats;
+
+import java.nio.ByteBuffer;
+import java.util.Map;
+import org.apache.iceberg.Schema;
+import org.apache.iceberg.expressions.Expression;
+import org.apache.iceberg.io.CloseableIterable;
+import org.apache.iceberg.mapping.NameMapping;
+
+/**
+ * Builder interface for creating file readers across supported data file formats. The {@link
+ * FormatModel} implementations provides appropriate {@link ReadBuilder} instances
+ *
+ * <p>The {@link ReadBuilder} follows the builder pattern to configure and create {@link
+ * CloseableIterable} instances that read data from source files. Configuration options include
+ * schema projection, predicate filtering, record batching, and encryption settings.
+ *
+ * <p>This interface is directly exposed to users for parameterizing readers.
+ */
+public interface ReadBuilder {
+  /**
+   * Restricts the read to the given range: [start, start + length).
+   *
+   * @param newStart the start position for this read
+   * @param newLength the length of the range this read should scan
+   */
+  ReadBuilder split(long newStart, long newLength);
+
+  /** Set the projection schema. */
+  ReadBuilder project(Schema schema);
+
+  /** Sets the expected output schema. If not provided derived from the {@link #project(Schema)}. */
+  default ReadBuilder outputSchema(Object schema) {
+    throw new UnsupportedOperationException("Not implemented yet");
+  }
+
+  /**
+   * Configures whether filtering should be case-sensitive. If the reader supports filtering, it
+   * must respect this setting.
+   *
+   * @param caseSensitive indicates if filtering is case-sensitive
+   */
+  ReadBuilder caseSensitive(boolean caseSensitive);
+
+  /**
+   * Pushes down the {@link Expression} filter for the reader to prevent reading unnecessary
+   * records. Some readers might not be able to filter some part of the exception. In this case the
+   * reader might return unfiltered or partially filtered rows. It is the caller's responsibility to
+   * apply the filter again. The default implementation sets the filter to be case-sensitive.
+   *
+   * @param filter the filter to set
+   */
+  ReadBuilder filter(Expression filter);
+
+  /**
+   * Sets configuration key/value pairs for the reader. Reader builders should ignore configuration
+   * keys not known for them.
+   */
+  ReadBuilder set(String key, String value);
+
+  /**
+   * Sets multiple configuration key/value pairs for the reader. Reader builders should ignore
+   * configuration keys not known for them.
+   */
+  default ReadBuilder setAll(Map<String, String> properties) {
+    properties.forEach(this::set);
+    return this;
+  }
+
+  /** Enables reusing the containers returned by the reader. Decreases pressure on GC. */
+  ReadBuilder reuseContainers();
+
+  /** Sets the batch size for vectorized readers. */
+  ReadBuilder recordsPerBatch(int numRowsPerBatch);
```

Comment:
This is another one where I know that's on the current ReadBuilder API but I wonder if it makes sense to  rename to be a more concise `batchSize` API?